File: src/main/java/org/stt/gui/jfx/text/PopupAtCaretPlacer.java
package org.stt.gui.jfx.text;

import javafx.geometry.Bounds;
import javafx.scene.Node;
import javafx.scene.control.TextArea;
import javafx.stage.Popup;
import javafx.stage.Window;

import java.util.Objects;

public class PopupAtCaretPlacer {
    private final Node path;
    private final Popup popup;
    private final Window window;

    public PopupAtCaretPlacer(final TextArea textArea, final Popup popup) {
        Objects.requireNonNull(textArea);
        this.popup = Objects.requireNonNull(popup);
        path = textArea.lookup("Path");
        this.window = textArea.getScene().getWindow();

        path.boundsInLocalProperty().addListener((observable, oldValue, newValue) -> updatePopupLocation(newValue));
        updatePopupLocation(path.getBoundsInLocal());
    }

    private void updatePopupLocation(Bounds bounds) {
        if (popup.isShowing()) {
            Bounds caretPosition = path.localToScene(bounds);
            double x = window.getX() + window.getScene().getX();
            double y = window.getY() + window.getScene().getY();
            popup.setX(caretPosition.getMaxX() + x);
            popup.setY(caretPosition.getMaxY() + y);
        }
    }
}

File: src/main/kotlin/org/stt/gui/jfx/text/PopupAtCaretPlacer.kt
package org.stt.gui.jfx.text

import javafx.geometry.Bounds
import javafx.scene.Node
import javafx.scene.control.TextArea
import javafx.stage.Popup
import javafx.stage.Window

class PopupAtCaretPlacer(textArea: TextArea, private val popup: Popup) {
    private val path: Node = textArea.lookup("Path")
    private val window: Window

    init {
        this.window = textArea.scene.window

        path.boundsInLocalProperty().addListener { _, _, newValue -> updatePopupLocation(newValue) }
        updatePopupLocation(path.boundsInLocal)
    }

    private fun updatePopupLocation(bounds: Bounds) {
        if (popup.isShowing) {
            val caretPosition = path.localToScene(bounds)
            val x = window.x + window.scene.x
            val y = window.y + window.scene.y
            popup.x = caretPosition.maxX + x
            popup.y = caretPosition.maxY + y
        }
    }
}
